group/assignment6/ParkingLot/ParkingLot.java

```diff
@@ -155,6 +159,71 @@ private void initStateMaps(Map<Car, RearrangeState> carStateMap, Map<Space, Spac
     public void moveCarToEmptySpace(Car car) {
         moveCarToEmptySpace(car,null);
     }
+    
+    
+    
+    
+    // Solution to challenge 1 and 2 ===============================================================
+
+    /**
+     * Finds a list of moves to rearrange cars from an initial state to a goal state.
+     * This method assumes there is exactly one empty space.
+     * @param initialState a ParkingLot with exactly one empty space
+     * @param goalState    a rearrangement of the initial ParkingLot
+     * @return a list of moves
+     */
+    public List<Move> getCarMoves(ParkingLot goalState) {
+      if (!isRearrangementOfParkingLot(goalState)) {
+        throw new IllegalArgumentException(
+            "The goal parking lot state has to be a rearrangement of the initial parking lot state.");
+      }
+
+      List<Move> moves = new ArrayList<Move>();
+      ParkingLot currentState = new ParkingLot(carSpaceBiMap.getKeyToValueMap());
+      moveCars(currentState, goalState, moves);
+      return moves;
+    }
+
+    private static void moveCars(ParkingLot currentState, ParkingLot goalState, List<Move> moves) {

```

Comment:
Nice! This is a clear and simple algorithm and I think we need to use it for challenge 1 and 2.